crates/ln-dlc-storage/src/lib.rs

```diff
@@ -0,0 +1,1191 @@
+        {
+            Some(res) => Ok(Some(deserialize_contract(&res.1)?)),
+            None => Ok(None),
+        }
+    }
+
+    fn get_contracts(&self) -> Result<Vec<Contract>, Error> {
+        let contracts = self.store.read(CONTRACT, None).map_err(to_storage_error)?;
+
+        let contracts = contracts
+            .iter()
+            .filter_map(|x| match deserialize_contract(&x.1) {
+                Ok(contract) => Some(contract),
+                Err(e) => {
+                    log::error!("Failed to deserialize contract: {e}");
+                    None
+                }
+            })
+            .collect();
+
+        Ok(contracts)
+    }
+
+    fn create_contract(&self, contract: &OfferedContract) -> Result<(), Error> {
+        let serialized = serialize_contract(&Contract::Offered(contract.clone()))?;
+        self.store
+            .write(CONTRACT, contract.id.to_vec(), serialized)
+            .map_err(to_storage_error)
+    }
+
+    fn delete_contract(&self, contract_id: &ContractId) -> Result<(), Error> {
+        self.store
+            .delete(CONTRACT, Some(contract_id.to_vec()))
+            .map_err(to_storage_error)
+    }
+
+    fn update_contract(&self, contract: &Contract) -> Result<(), Error> {
+        let serialized = serialize_contract(contract)?;
+
+        match contract {
+            a @ Contract::Accepted(_) | a @ Contract::Signed(_) => {
+                self.store
+                    .delete(CONTRACT, Some(a.get_temporary_id().to_vec()))
+                    .map_err(to_storage_error)?;
+            }
+            _ => {}
+        };
+
+        self.store
+            .write(CONTRACT, contract.get_id().to_vec(), serialized)
+            .map_err(to_storage_error)
+    }
+
+    fn get_contract_offers(&self) -> Result<Vec<OfferedContract>, Error> {
+        let contracts = self.get_raw_contracts()?;
+
+        self.get_data_with_prefix(&contracts, &[ContractPrefix::Offered.into()], None)
+    }
+
+    fn get_signed_contracts(&self) -> Result<Vec<SignedContract>, Error> {
+        let contracts = self.get_raw_contracts()?;
+
+        self.get_data_with_prefix(&contracts, &[ContractPrefix::Signed.into()], None)
+    }
+
+    fn get_confirmed_contracts(&self) -> Result<Vec<SignedContract>, Error> {
+        let contracts = self.get_raw_contracts()?;
+
+        self.get_data_with_prefix(&contracts, &[ContractPrefix::Confirmed.into()], None)
+    }
+
+    fn get_preclosed_contracts(&self) -> Result<Vec<PreClosedContract>, Error> {
+        let contracts = self.get_raw_contracts()?;
+
+        self.get_data_with_prefix(&contracts, &[ContractPrefix::PreClosed.into()], None)
+    }
+
+    fn upsert_channel(&self, channel: Channel, contract: Option<Contract>) -> Result<(), Error> {
+        let serialized = serialize_channel(&channel)?;
+        let serialized_contract = match contract.as_ref() {
+            Some(c) => Some(serialize_contract(c)?),
+            None => None,
+        };
```

Comment:
That's code I've actually copied. I guess its not a map because the `serialize_contract` returns a result and that can't be handled equally in a map function.